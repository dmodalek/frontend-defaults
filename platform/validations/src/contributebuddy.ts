import { ContributebuddyAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
	ValidationResult,
	ValidationSeverityLevel,
	validateInstallation,
} from '@namics/frontend-defaults-platform-core';

export const contributebuddyValidation = async (
	cwd: string,
	analytics: ContributebuddyAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.contributebuddy) {
		// contributebuddy is used in the project, validation below
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.contributebuddyInstallation,
				name: 'Contributebuddy',
				source: 'contributebuddyValidation',
			})
		);

		if (!analytics.contributebuddyConfigFolder) {
			validationResults.push({
				level: ValidationSeverityLevel.error,
				message: 'No .contributebuddy folder found, make sure you set-up correctly',
				source: 'contributebuddyValidation',
			});
		}

		return validationResults;
	}

	return [
		{
			level: ValidationSeverityLevel.info,
			message: 'We recommend to use contribute-buddy to obtain maintainability',
			source: 'contributebuddyValidation',
		},
	];
};
