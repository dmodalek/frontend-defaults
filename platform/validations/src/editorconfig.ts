import { EditorconfigAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const editorconfigValidation = async (cwd: string, analytics: EditorconfigAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.editorConfig) {
		// TODO: editorconfig is available, so we need to validate all things
		return [];
	}

	return [{
		level: ValidationSeverityLevel.error,
		message: 'Default namics editorconfig file is required in every project',
		source: 'editorconfigValidation'
	}];
}