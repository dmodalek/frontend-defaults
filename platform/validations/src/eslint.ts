import { ESLintAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

/**
 * TODO: Should we check if a project is a TS project in advance? So we wouldn't need to run this validation atm.
 * I think we'll need to update this as soon as TSLint will be integrated into ESLint
 */
export const eslintValidation = async (cwd: string, analytics: ESLintAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.eslint) {
		// TODO: eslint is available, so we need to validate all things
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.eslintInstallation,
				name: 'ESLint',
				source: 'eslintValidation'
			})
		);

		if (!analytics.eslintIgnore) {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'No .eslintignore found',
				source: 'eslintValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.warning,
		message: 'ESLint configuration missing',
		source: 'eslintValidation'
	}];
}