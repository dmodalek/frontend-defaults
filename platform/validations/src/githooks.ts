import { GitHooksAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

export const gitHooksValidation = async (cwd: string, analytics: GitHooksAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.gitHooks) {
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.gitHooksInstallation,
				name: 'GitHooks',
				source: 'gitHooksValidation'
			})
		);

		if (analytics.gitHooksLocation === 'rc') {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'For future projects we decided to use the package.json for configuration (.rc file used)',
				source: 'gitHooksValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.info,
		message: 'GitHooks are missing',
		source: 'gitHooksValidation'
	}];
}