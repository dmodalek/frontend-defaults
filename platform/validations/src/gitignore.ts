import { GitIgnoreAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const gitIgnoreValidation = async (
	cwd: string,
	analytics: GitIgnoreAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.gitignore) {
		// TODO: gitignore is available, so we need to validate the contents
		return [];
	}

	return [
		{
			level: ValidationSeverityLevel.error,
			message: 'Git ignore missing',
			source: 'gitIgnoreValidation',
		},
	];
};
