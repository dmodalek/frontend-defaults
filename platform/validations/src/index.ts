export * from './editorconfig';
export * from './eslint';
export * from './githooks';
export * from './gitignore';
export * from './license';
export * from './nitro';
export * from './node';
export * from '../../patches/src/npmrc';
export * from './readme';
export * from './tslint';
export * from './typescript';
export * from './webpack';
