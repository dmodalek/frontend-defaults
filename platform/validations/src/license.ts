import { LicenseAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const licenseValidation = async (cwd: string, analytics: LicenseAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.license) {
		// TODO: license is available, so we need to validate the content
		return [];
	}

	return [{
		level: ValidationSeverityLevel.warning,
		message: 'You don\'t have a license file in your project',
		source: 'licenseValidation'
	}];
}