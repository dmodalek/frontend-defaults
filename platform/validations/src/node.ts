import { NodeAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
    checkNewerLTSVersion,
    ValidationException,
    ValidationResult,
    ValidationSeverityLevel
    } from '@namics/frontend-defaults-platform-core';

export const nodeValidation = async (cwd: string, analytics: NodeAnalyzerResult): Promise<ValidationResult[]> => {
    const { nodeVersion, nodeVersionManagerInfo } = analytics;

    if (!nodeVersion || !nodeVersionManagerInfo) {
        return [
            {
                message: `No node version manager used, integrate node-version or/and nvm`,
                level: ValidationSeverityLevel.error,
                source: 'NodeValidation',
                patch: ['NVMInstallPatch', 'NodeVersionInstallPatch']
            }
        ];
    }

    const validations: ValidationResult[] = [];

    if (!nodeVersionManagerInfo.nvm) {
        // no .nvmrc file found
        validations.push({
            message: `No .nvmrc ressource file found, you might add one`,
            source: 'NodeValidation',
            level: ValidationSeverityLevel.info,
            patch: ['NVMInstallPatch']
        });
    } else if (typeof nodeVersionManagerInfo.nvm === 'string') {
        // check for lts updates in .nvmrc
        const nvmUpgradeException = await checkManagerForUpdate(nodeVersionManagerInfo.nvm, 'nvm', 'UpdateNVMPatch');
        if (nvmUpgradeException) {
            validations.push(nvmUpgradeException);
        }
    }

    if (!nodeVersionManagerInfo['node-version']) {
        // no .node-version file found
        validations.push({
            message: `No .node-version ressource file found, you might add one`,
            source: 'NodeValidation',
            level: ValidationSeverityLevel.warning,
            patch: ['NodeVersionInstallPatch']
        });
    } else if (typeof nodeVersionManagerInfo["node-version"] === 'string') {
        // check for lts updates in .node-version
        const nodeVersionUpgradeException = await checkManagerForUpdate(nodeVersionManagerInfo["node-version"], 'node-version', 'UpdateNodeVersionPatch');
        if (nodeVersionUpgradeException) {
            validations.push(nodeVersionUpgradeException);
        }
    }

    return validations;
}

/**
 * Check if there's an LTS update for a certain manager
 * @param selectedVersion current node version from file
 * @param config configuration type (nvm or node-version)
 * @param patch patch needed if upgrade available
 */
const checkManagerForUpdate = async (selectedVersion: string, config: string, patch: string): Promise<ValidationResult | void> => {
    const checkForUpdate = await checkNewerLTSVersion('node', selectedVersion);

    if (checkForUpdate.upgradable) {
        return {
            message: `There's a newer node LTS version available for ${config}, try upgrading`,
            source: 'NodeValidation',
            level: ValidationSeverityLevel.info,
            patch: [
                [patch, {
                    current: checkForUpdate.current,
                    latest: checkForUpdate.latest
                }]
            ]
        }
    }
}