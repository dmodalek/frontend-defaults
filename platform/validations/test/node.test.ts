import { checkInternetConnection, getFixtureDir, getLatestVersion } from './utils';
import { nodeValidation } from '../src/node';
import { nodeAnalyzer, NodeAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { analyze } from '@namics/frontend-defaults-platform-core';

const generateAnalyticsFor = async (fixtureContext: string): Promise<NodeAnalyzerResult> => {
    return await analyze(fixtureContext, nodeAnalyzer);
};

const FIXTURE_VERSION_MANAGER = getFixtureDir('version-manager-project');
const FIXTURE_NO_MANAGER = getFixtureDir('default-project');
const FIXTURE_NVM = getFixtureDir('nvm-only-project');
const FIXTURE_NODE_VERSION = getFixtureDir('node-version-only-project');

beforeAll(async () => {
    if (!(await checkInternetConnection())) {
        console.debug(`Please only run tests while connected to the internet!`);
        process.exit(1); // kill child worker
    }
});

describe('Validations', () => {
    describe('NodeValidations', () => {
        it('should not thorw any error', async () => {
            expect(async () => {
                const analytics = await generateAnalyticsFor(FIXTURE_VERSION_MANAGER);
                nodeValidation(FIXTURE_VERSION_MANAGER, analytics);
            }).not.toThrow();
        });

        it('should recognize need to install both managers', async () => {
            const analytics = await generateAnalyticsFor(FIXTURE_NO_MANAGER);
            const exceptions = await nodeValidation(FIXTURE_NO_MANAGER, analytics);


            expect(exceptions).toHaveLength(1);
            expect(exceptions).toMatchSnapshot();
            expect(exceptions[0].patch).toEqual(['NVMInstallPatch', 'NodeVersionInstallPatch']);
        });

        it('should recognize both managers and their updates', async () => {
            const analytics = await generateAnalyticsFor(FIXTURE_VERSION_MANAGER);
            const exceptions = await nodeValidation(FIXTURE_VERSION_MANAGER, analytics);

            expect(exceptions).toHaveLength(2);
            expect(exceptions).toMatchSnapshot();

            const [nvmUpdate, nodeVersionUpdate] = exceptions;
            expect(nvmUpdate.patch![0]).toEqual(['UpdateNVMPatch', {
                current: '9.0.0',
                latest: await getLatestVersion('node', true)
            }]);
            expect(nodeVersionUpdate.patch![0]).toEqual(['UpdateNodeVersionPatch', {
                current: '9.0.0',
                latest: await getLatestVersion('node', true)
            }]);
        });

        it('should recognize the nvm manager, nvm update and node-version installation', async () => {
            const analytics = await generateAnalyticsFor(FIXTURE_NVM);
            const exceptions = await nodeValidation(FIXTURE_VERSION_MANAGER, analytics);

            expect(exceptions).toHaveLength(2);
            expect(exceptions).toMatchSnapshot();

            const [nvmUpdate, nodeVersionInstall] = exceptions;
            expect(nvmUpdate.patch![0]).toEqual(['UpdateNVMPatch', {
                current: '10.0.0',
                latest: await getLatestVersion('node', true)
            }]);
            expect(nodeVersionInstall.patch![0]).toEqual('NodeVersionInstallPatch');
        });

        it('should recognize the node-version manager, node-version update and nvm installation', async () => {
            const analytics = await generateAnalyticsFor(FIXTURE_NODE_VERSION);
            const exceptions = await nodeValidation(FIXTURE_VERSION_MANAGER, analytics);

            expect(exceptions).toHaveLength(2);
            expect(exceptions).toMatchSnapshot();

            const [nvmInstall, nodeVersionUpdate] = exceptions;
            expect(nvmInstall.patch![0]).toEqual('NVMInstallPatch');
            expect(nodeVersionUpdate.patch![0]).toEqual(['UpdateNodeVersionPatch', {
                current: '10.0.0',
                latest: await getLatestVersion('node', true)
            }]);
        });
    });
});