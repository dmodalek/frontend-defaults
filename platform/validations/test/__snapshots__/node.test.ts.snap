// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validations NodeValidations should recognize both managers and their updates 1`] = `
Array [
  Object {
    "level": "INFO",
    "message": "There's a newer node LTS version available for nvm, try upgrading",
    "patch": Array [
      Array [
        "UpdateNVMPatch",
        Object {
          "current": "9.0.0",
          "latest": "10.15.1",
        },
      ],
    ],
    "source": "NodeValidation",
  },
  Object {
    "level": "INFO",
    "message": "There's a newer node LTS version available for node-version, try upgrading",
    "patch": Array [
      Array [
        "UpdateNodeVersionPatch",
        Object {
          "current": "9.0.0",
          "latest": "10.15.1",
        },
      ],
    ],
    "source": "NodeValidation",
  },
]
`;

exports[`Validations NodeValidations should recognize need to install both managers 1`] = `
Array [
  Object {
    "level": "ERROR",
    "message": "No node version manager used, integrate node-version or/and nvm",
    "patch": Array [
      "NVMInstallPatch",
      "NodeVersionInstallPatch",
    ],
    "source": "NodeValidation",
  },
]
`;

exports[`Validations NodeValidations should recognize the node-version manager, node-version update and nvm installation 1`] = `
Array [
  Object {
    "level": "INFO",
    "message": "No .nvmrc ressource file found, you might add one",
    "patch": Array [
      "NVMInstallPatch",
    ],
    "source": "NodeValidation",
  },
  Object {
    "level": "INFO",
    "message": "There's a newer node LTS version available for node-version, try upgrading",
    "patch": Array [
      Array [
        "UpdateNodeVersionPatch",
        Object {
          "current": "10.0.0",
          "latest": "10.15.1",
        },
      ],
    ],
    "source": "NodeValidation",
  },
]
`;

exports[`Validations NodeValidations should recognize the nvm manager, nvm update and node-version installation 1`] = `
Array [
  Object {
    "level": "INFO",
    "message": "There's a newer node LTS version available for nvm, try upgrading",
    "patch": Array [
      Array [
        "UpdateNVMPatch",
        Object {
          "current": "10.0.0",
          "latest": "10.15.1",
        },
      ],
    ],
    "source": "NodeValidation",
  },
  Object {
    "level": "WARNING",
    "message": "No .node-version ressource file found, you might add one",
    "patch": Array [
      "NodeVersionInstallPatch",
    ],
    "source": "NodeValidation",
  },
]
`;
